README.md
# pasystral
pasystrail with patches for GalliumOS


debian/changelog
pasystray (0.6.0-1ubuntu1-galliumos0) bismuth; urgency=medium

  * Prepare for GalliumOS 3.0
    - patches from 0.4.0-1galliumos4

 -- reynhout <reynhout@galliumos.org>  Sun, 06 May 2018 19:09:18 +0000

pasystray (0.6.0-1ubuntu1) yakkety; urgency=medium

  * Merge from Debian unstable. Remaining changes:
    - debian/control: Build-depend on libappindicator3-dev for indicator
      support.

 -- Logan Rosen <logan@ubuntu.com>  Fri, 22 Jul 2016 23:11:26 -0400

pasystray (0.6.0-1) unstable; urgency=low

  * New upstream release 0.6.0.
  * Update debian/copyright.
  * Bump Standards-Version to 3.9.8.

 -- Scott Leggett <scott@sl.id.au>  Mon, 18 Jul 2016 23:19:07 +1000

pasystray (0.5.2-1ubuntu1) yakkety; urgency=medium

  * d/control: Build-dep on libappindicator3-dev for indicator support.

 -- Unit 193 <unit193@ubuntu.com>  Thu, 02 Jun 2016 21:24:38 -0400

pasystray (0.5.2-1) unstable; urgency=low

  * New upstream release 0.5.2 (Closes: #792461)
  * Remove .desktop file patch applied upstream.
  * Update package VCS URLs.
  * Bump Standards-Version from 3.9.4 to 3.9.7.
  * Enable all hardening build options.
  * Add "Suggests" dependency on pavucontrol, and expand long description.
    (Closes: #792459)
  * Remove manpage from package (accepted upstream).
  * Remove Debian menu file in favour of .desktop file.
  * Update copyright holder information.

 -- Scott Leggett <scott@sl.id.au>  Wed, 27 Apr 2016 23:12:20 +1000

pasystray (0.4.0-1galliumos4) xenial; urgency=low

  * Put GalliumOS changes into patches
  * New package for xenial

 -- Hugh Greenberg <hugh@galliumos.org>  Wed, 16 Mar 2016 21:05:33 +0000

pasystray (0.4.0-1galliumos3) vivid-galliumos; urgency=low

  * Set package version to match provenance

 -- reynhout <reynhout@galliumos.org>  Sun, 10 Jan 2016 21:05:33 +0000

pasystray (0.4.0.3-2) unstable; urgency=low

  * Mods for GalliumOS

 -- Hugh Greenberg <hugh@galliumos>  Sun, 29 Nov 2015 18:00:59 +1100

pasystray (0.4.0-1) unstable; urgency=low

  * Initial release (Closes: #729205)

 -- Scott Leggett <scott@sl.id.au>  Sun, 10 Nov 2013 18:00:59 +1100



src/ui.h
/***
  This file is part of PaSystray

  Copyright (C) 2011, 2012 Christoph Gysin

  PaSystray is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  PaSystray is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with PaSystray; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifndef PASYSTRAY_UI_H
#define PASYSTRAY_UI_H

#include "menu_info.h"
#include <gtk/gtk.h>

void ui_load();
void ui_update_systray_icon(menu_info_item_t* mii);
GtkAboutDialog* ui_aboutdialog();

GtkDialog* ui_renamedialog();
GtkLabel* ui_renamedialog_label();
GtkEntry* ui_renamedialog_entry();

#endif /* PASYSTRAY_UI_H */


src/options.c
/***
  This file is part of PaSystray

  Copyright (C) 2011, 2012 Christoph Gysin

  PaSystray is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  PaSystray is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with PaSystray; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#include <stdlib.h>
#include <glib.h>
#include <gtk/gtk.h>
#include "config.h"

static gboolean version = FALSE;

static GOptionEntry entries[] =
{
    { "version", 'V', 0, G_OPTION_ARG_NONE, &version, "print version and exit", NULL },
    { .long_name = NULL }
};

void parse_options(int argc, char *argv[])
{
    GError *error = NULL;
    GOptionContext *context;

    context = g_option_context_new(NULL);
    g_option_context_add_main_entries(context, entries, NULL);
    g_option_context_add_group(context, gtk_get_option_group(TRUE));
    if(!g_option_context_parse(context, &argc, &argv, &error))
    {
        g_print("option parsing failed: %s\n", error->message);
        g_error_free(error);
        exit(1);
    }

    if(version)
    {
        g_print("%s\n", VERSION);
        exit(1);
    }
}


src/pasystray.h
/***
  This file is part of PaSystray

  Copyright (C) 2011, 2012 Christoph Gysin

  PaSystray is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  PaSystray is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with PaSystray; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifndef PASYSTRAY_PASYSTRAY_H
#define PASYSTRAY_PASYSTRAY_H

void init();
void quit();
void destroy();

#endif /* PASYSTRAY_PASYSTRAY_H */


src/pulseaudio_action.c
/***
  This file is part of PaSystray

  Copyright (C) 2011, 2012 Christoph Gysin

  PaSystray is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  PaSystray is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with PaSystray; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#include "pulseaudio_action.h"

#include <pulse/volume.h>
#include <pulse/ext-device-manager.h>

#include "notify.h"
#include "x11-property.h"

extern pa_context* context;

void pulseaudio_set_default(menu_info_item_t* mii)
{
    pa_operation* o = NULL;

    switch(mii->menu_info->type)
    {
        case MENU_SERVER:
            if(mii->index == 0)
                x11_property_del("PULSE_SERVER");
            else
                x11_property_set("PULSE_SERVER", mii->address);
            break;
        case MENU_SINK:
            o = pa_context_set_default_sink(context, mii->name,
                    pulseaudio_set_default_success_cb, mii);
            break;
        case MENU_SOURCE:
            o = pa_context_set_default_source(context, mii->name,
                    pulseaudio_set_default_success_cb, mii);
            break;
        case MENU_INPUT:
        case MENU_OUTPUT:
            /* nothing to do here */
            break;
    }
    if(o)
        pa_operation_unref(o);
}

void pulseaudio_set_default_success_cb(pa_context *c, int success, void *userdata)
{
    menu_info_item_t* mii = userdata;

    if(!success)
        g_error("failed to set default to %s \"%s\"!\n",
                menu_info_type_name(mii->menu_info->type), mii->name);
}

void pulseaudio_move_input_to_sink(menu_info_item_t* input, menu_info_item_t* sink)
{
    pa_operation_unref(pa_context_move_sink_input_by_index(context, input->index,
                sink->index, pulseaudio_move_success_cb, input));
}

void pulseaudio_move_output_to_source(menu_info_item_t* output, menu_info_item_t* source)
{
    pa_operation_unref(pa_context_move_source_output_by_index(context, output->index,
                source->index, pulseaudio_move_success_cb, output));
}

void pulseaudio_move_success_cb(pa_context *c, int success, void *userdata)
{
    menu_info_item_t* to = userdata;
    menu_info_item_t* from = to->menu_info->parent;

    if(!success)
        g_error("failed to move %s '%s' to %s '%s'!\n",
                menu_info_type_name(from->menu_info->type), from->name,
                menu_info_type_name(to->menu_info->type), to->name);
}

void pulseaudio_rename(menu_info_item_t* mii, const char* name)
{
#ifdef DEBUG
    g_message("rename %s '%s' to '%s'",
            menu_info_type_name(mii->menu_info->type), mii->desc, name);
#endif
    char *key = g_markup_printf_escaped("%s:%s", menu_info_type_name(mii->menu_info->type), mii->name);

    pa_operation* o;
    if(!(o = pa_ext_device_manager_set_device_description(context, key, name, pulseaudio_rename_success_cb, mii))) {
        g_error("pa_ext_device_manager_set_device_description(context, %s, %s) failed", key, name);
        return;
    }
    pa_operation_unref(o);
}

void pulseaudio_rename_success_cb(pa_context *c, int success, void *userdata)
{
    menu_info_item_t* mii = userdata;

    // TODO: try to autoload module-device-manager?
    if(!success)
        g_warning("failed to rename %s '%s'! module-device-manager loaded?\n",
                menu_info_type_name(mii->menu_info->type), mii->name);
}

void pulseaudio_volume(menu_info_item_t* mii, int inc)
{
#ifdef DEBUG
    g_message("pulseaudio_volume(%s, %i)", mii->name, inc);
#endif

    /* increment/decrement in 2% steps */
    pa_cvolume* volume;
    if(inc < 0)
        volume = pa_cvolume_dec(mii->volume, -inc * PA_VOLUME_NORM / 50);
    else if(inc > 0) {
	volume = pa_cvolume_inc_clamp(mii->volume, inc * PA_VOLUME_NORM / 50,
				      PA_VOLUME_NORM * 100 / 100);
    } else
        return;

    pa_operation* o = NULL;

    switch(mii->menu_info->type)
    {
        case MENU_SERVER:
            /* nothing to do here */
            break;
        case MENU_SINK:
            o = pa_context_set_sink_volume_by_index(context, mii->index,
                    volume, pulseaudio_set_volume_success_cb, mii);
            break;
        case MENU_SOURCE:
            o = pa_context_set_source_volume_by_index(context, mii->index,
                    volume, pulseaudio_set_volume_success_cb, mii);
            break;
        case MENU_INPUT:
            o = pa_context_set_sink_input_volume(context, mii->index,
                    volume, pulseaudio_set_volume_success_cb, mii);
            break;
        case MENU_OUTPUT:
            o = pa_context_set_source_output_volume(context, mii->index,
                    volume, pulseaudio_set_volume_success_cb, mii);
            break;
    }
    if(o)
        pa_operation_unref(o);
}

void pulseaudio_set_volume_success_cb(pa_context *c, int success, void *userdata)
{
    menu_info_item_t* mii = userdata;

    if(!success)
    {
        g_error("failed to set volume for %s \"%s\"!\n",
                menu_info_type_name(mii->menu_info->type), mii->name);
        return;
    }
}

void pulseaudio_update_volume_notification(menu_info_item_t* mii)
{
    char vol[PA_CVOLUME_SNPRINT_MAX];
    gchar* msg = g_strdup_printf("%s %s: %s%s",
                menu_info_type_name(mii->menu_info->type), mii->desc,
                pa_cvolume_snprint(vol, sizeof(vol), mii->volume),
                mii->mute ? " [muted]" : "");

    if(!mii->notify)
        mii->notify = notify(msg, NULL, mii->icon);
    else
        notify_update(mii->notify, msg, NULL, mii->icon);

    g_free(msg);
}

void pulseaudio_toggle_mute(menu_info_item_t* mii)
{
#ifdef DEBUG
    g_message("pulseaudio_toggle_mute(%s)", mii->name);
#endif

    pa_operation* o = NULL;

    int mute = (mii->mute) ? 0 : 1;

    switch(mii->menu_info->type)
    {
        case MENU_SERVER:
            /* nothing to do here */
            break;
        case MENU_SINK:
            o = pa_context_set_sink_mute_by_index(context, mii->index,
                    mute, pulseaudio_toggle_mute_success_cb, mii);
            break;
        case MENU_SOURCE:
            o = pa_context_set_source_mute_by_index(context, mii->index,
                    mute, pulseaudio_toggle_mute_success_cb, mii);
            break;
        case MENU_INPUT:
            o = pa_context_set_sink_input_mute(context, mii->index,
                    mute, pulseaudio_toggle_mute_success_cb, mii);
            break;
        case MENU_OUTPUT:
            o = pa_context_set_source_output_mute(context, mii->index,
                    mute, pulseaudio_toggle_mute_success_cb, mii);
            break;
    }
    if(o)
        pa_operation_unref(o);
}

void pulseaudio_toggle_mute_success_cb(pa_context *c, int success, void *userdata)
{
    menu_info_item_t* mii = userdata;

    if(!success)
        g_error("failed to toogle mute for %s \"%s\"!\n",
                menu_info_type_name(mii->menu_info->type), mii->name);
}


src/pulseaudio.c
/***
  This file is part of PaSystray

  Copyright (C) 2011, 2012 Christoph Gysin

  PaSystray is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  PaSystray is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with PaSystray; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#include "pulseaudio.h"
#include "pulseaudio_info.h"
#include "systray.h"
#include "systray_impl.h"
#include "notify.h"

#include <pulse/glib-mainloop.h>

pa_context* context = NULL;

static pa_glib_mainloop* m = NULL;
static pa_proplist* context_proplist = NULL;
static char* server = NULL;

void pulseaudio_init(menu_infos_t* mis)
{
    if(!(m = pa_glib_mainloop_new(g_main_context_default())))
        pulseaudio_quit("pa_glib_mainloop_new() failed.");

    pulseaudio_prepare_context();
    pa_context_set_state_callback(context, pulseaudio_context_state_cb, mis);

    pulseaudio_connect();
}

void pulseaudio_destroy()
{
    pulseaudio_quit(NULL);
}

void pulseaudio_prepare_context()
{
    pa_mainloop_api* mainloop_api = pa_glib_mainloop_get_api(m);

    context_proplist = pa_proplist_new();
    pa_proplist_sets(context_proplist, PA_PROP_APPLICATION_NAME, "PulseAudio systray");

    context = pa_context_new_with_proplist(mainloop_api, NULL, context_proplist);
    if(!context)
        pulseaudio_quit("pa_context_new() failed.");
}

void pulseaudio_connect()
{
    if(pa_context_connect(context, server, 0, NULL) < 0)
    {
        g_message("pa_context_connect() failed: ");
        pulseaudio_quit(pa_strerror(pa_context_errno(context)));
    }
}

void pulseaudio_context_state_cb(pa_context* c, void* userdata)
{
    menu_infos_t* mis = userdata;

    switch(pa_context_get_state(c))
    {
        case PA_CONTEXT_UNCONNECTED:
            g_message("PulseAudio context unconnected!");
            break;

        case PA_CONTEXT_READY:
        {
	    char* tooltip = g_strdup("Sound Control");
            systray_impl_set_tooltip(mis->systray, tooltip);
            g_free(tooltip);

            pa_context_set_subscribe_callback(context, pulseaudio_event_cb, mis);
            pa_operation_unref(pa_context_subscribe(context,
                        PA_SUBSCRIPTION_MASK_SINK |
                        PA_SUBSCRIPTION_MASK_SOURCE |
                        PA_SUBSCRIPTION_MASK_SINK_INPUT |
                        PA_SUBSCRIPTION_MASK_SOURCE_OUTPUT |
                        PA_SUBSCRIPTION_MASK_SERVER,
                        pulseaudio_subscribed_cb, mis));
            break;
        }

        case PA_CONTEXT_FAILED:
        case PA_CONTEXT_TERMINATED:
            g_message("PulseAudio terminated!");
            menu_infos_clear(mis);
            pa_context_unref(context);
            pulseaudio_prepare_context();
            g_message("reconnecting...");
            pulseaudio_connect();
            break;

        case PA_CONTEXT_CONNECTING:
        case PA_CONTEXT_AUTHORIZING:
        case PA_CONTEXT_SETTING_NAME:
            break;
    }
}

void pulseaudio_subscribed_cb(pa_context* c, int success, void* userdata)
{
    menu_infos_t* mis = userdata;
    pa_operation_unref(pa_context_get_server_info(context,
                pulseaudio_server_init_cb, &mis->menu_info[MENU_SERVER]));
    pa_operation_unref(pa_context_get_sink_info_list(context,
                pulseaudio_sink_init_cb, &mis->menu_info[MENU_SINK]));
    pa_operation_unref(pa_context_get_source_info_list(context,
                pulseaudio_source_init_cb, &mis->menu_info[MENU_SOURCE]));
    pa_operation_unref(pa_context_get_sink_input_info_list(context,
                pulseaudio_sink_input_init_cb, &mis->menu_info[MENU_INPUT]));
    pa_operation_unref(pa_context_get_source_output_info_list(context,
                pulseaudio_source_output_init_cb, &mis->menu_info[MENU_OUTPUT]));
}

void pulseaudio_event_cb(pa_context* c, pa_subscription_event_type_t t, uint32_t index, void* userdata)
{
#ifdef DEBUG
    pulseaudio_print_event(t, index);
#endif

    menu_infos_t* mis = userdata;

    pa_subscription_event_type_t type = t & PA_SUBSCRIPTION_EVENT_TYPE_MASK;
    pa_subscription_event_type_t facility = t & PA_SUBSCRIPTION_EVENT_FACILITY_MASK;

    switch(type)
    {
        case PA_SUBSCRIPTION_EVENT_NEW:
            pulseaudio_event_new(facility, index, mis);
            break;
        case PA_SUBSCRIPTION_EVENT_CHANGE:
            pulseaudio_event_change(facility, index, mis);
            break;
        case PA_SUBSCRIPTION_EVENT_REMOVE:
            pulseaudio_event_remove(facility, index, mis);
            break;
        default:
            break;
    }
}

void pulseaudio_event_new(pa_subscription_event_type_t facility, uint32_t index, menu_infos_t* mis)
{
    switch(facility)
    {
        case PA_SUBSCRIPTION_EVENT_SERVER:
            pa_operation_unref(pa_context_get_server_info(context,
                        pulseaudio_server_add_cb, &mis->menu_info[MENU_SERVER]));
            break;
        case PA_SUBSCRIPTION_EVENT_SINK:
            pa_operation_unref(pa_context_get_sink_info_by_index(context, index,
                        pulseaudio_sink_add_cb, &mis->menu_info[MENU_SINK]));
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE:
            pa_operation_unref(pa_context_get_source_info_by_index(context, index,
                        pulseaudio_source_add_cb, &mis->menu_info[MENU_SOURCE]));
            break;
        case PA_SUBSCRIPTION_EVENT_SINK_INPUT:
            pa_operation_unref(pa_context_get_sink_input_info(context, index,
                        pulseaudio_sink_input_add_cb, &mis->menu_info[MENU_INPUT]));
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE_OUTPUT:
            pa_operation_unref(pa_context_get_source_output_info(context, index,
                        pulseaudio_source_output_add_cb, &mis->menu_info[MENU_OUTPUT]));
            break;
        default:
            break;
    }
}

void pulseaudio_event_change(pa_subscription_event_type_t facility, uint32_t index, menu_infos_t* mis)
{
    switch(facility)
    {
        case PA_SUBSCRIPTION_EVENT_SERVER:
            pa_operation_unref(pa_context_get_server_info(context,
                        pulseaudio_server_change_cb, &mis->menu_info[MENU_SERVER]));
            break;
        case PA_SUBSCRIPTION_EVENT_SINK:
            pa_operation_unref(pa_context_get_sink_info_by_index(context, index,
                        pulseaudio_sink_change_cb, &mis->menu_info[MENU_SINK]));
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE:
            pa_operation_unref(pa_context_get_source_info_by_index(context, index,
                        pulseaudio_source_change_cb, &mis->menu_info[MENU_SOURCE]));
            break;
        case PA_SUBSCRIPTION_EVENT_SINK_INPUT:
            pa_operation_unref(pa_context_get_sink_input_info(context, index,
                        pulseaudio_sink_input_change_cb, &mis->menu_info[MENU_INPUT]));
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE_OUTPUT:
            pa_operation_unref(pa_context_get_source_output_info(context, index,
                        pulseaudio_source_output_change_cb, &mis->menu_info[MENU_OUTPUT]));
            break;
        default:
            break;
    }
}

void pulseaudio_event_remove(pa_subscription_event_type_t facility, uint32_t index, menu_infos_t* mis)
{
    switch(facility)
    {
        case PA_SUBSCRIPTION_EVENT_SERVER:
            menu_info_item_remove(&mis->menu_info[MENU_SERVER], index);
            break;
        case PA_SUBSCRIPTION_EVENT_SINK:
            menu_info_item_remove(&mis->menu_info[MENU_SINK], index);
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE:
            menu_info_item_remove(&mis->menu_info[MENU_SOURCE], index);
            break;
        case PA_SUBSCRIPTION_EVENT_SINK_INPUT:
            menu_info_item_remove(&mis->menu_info[MENU_INPUT], index);
            break;
        case PA_SUBSCRIPTION_EVENT_SOURCE_OUTPUT:
            menu_info_item_remove(&mis->menu_info[MENU_OUTPUT], index);
            break;
        default:
            break;
    }
}

#ifdef DEBUG
void pulseaudio_print_event(pa_subscription_event_type_t t, uint32_t index)
{
    pa_subscription_event_type_t type = t & PA_SUBSCRIPTION_EVENT_TYPE_MASK;
    pa_subscription_event_type_t facility = t & PA_SUBSCRIPTION_EVENT_FACILITY_MASK;

    g_message("event %s on %s (%u)",
        (type == PA_SUBSCRIPTION_EVENT_NEW) ? "new" :
        (type == PA_SUBSCRIPTION_EVENT_CHANGE) ? "change" :
        (type == PA_SUBSCRIPTION_EVENT_REMOVE) ? "remove" :
        "unknown",
        (facility == PA_SUBSCRIPTION_EVENT_SINK) ? "sink" :
        (facility == PA_SUBSCRIPTION_EVENT_SOURCE) ? "source" :
        (facility == PA_SUBSCRIPTION_EVENT_SINK_INPUT) ? "sink-input" :
        (facility == PA_SUBSCRIPTION_EVENT_SOURCE_OUTPUT) ? "source-output" :
        (facility == PA_SUBSCRIPTION_EVENT_SERVER) ? "server" :
        "unknown",
        index);
}
#endif

void pulseaudio_server_init_cb(pa_context* c, const pa_server_info* i, void* userdata)
{
    pulseaudio_server_add(i, userdata, FALSE);
}

void pulseaudio_server_add_cb(pa_context* c, const pa_server_info* i, void* userdata)
{
    pulseaudio_server_add(i, userdata, TRUE);
}

void pulseaudio_server_change_cb(pa_context* c, const pa_server_info* i, void* userdata)
{
    pulseaudio_server_add(i, userdata, FALSE);
}

void pulseaudio_server_add(const pa_server_info* i, void* userdata, gboolean is_new)
{
    menu_info_t* mi = userdata;
    char* tooltip = server_info_str(i);
    menu_info_item_update(mi, 0, "localhost", i->host_name, NULL, 0, tooltip, NULL, NULL, -1);
    g_free(tooltip);

    /* set default sink/source */
    pulseaudio_change_default_item(&mi->menu_infos->menu_info[MENU_SINK], i->default_sink_name);
    pulseaudio_change_default_item(&mi->menu_infos->menu_info[MENU_SOURCE], i->default_source_name);
}

void pulseaudio_change_default_item(menu_info_t* mi, const char* new_default)
{
    g_free(mi->default_name);
    mi->default_name = g_strdup(new_default);

    menu_info_item_t* item = menu_info_item_get_by_name(mi, new_default);

    if(item)
        gtk_check_menu_item_set_active(GTK_CHECK_MENU_ITEM(item->widget), TRUE);
}

void pulseaudio_sink_init_cb(pa_context* c, const pa_sink_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_add(i, is_last, userdata, FALSE);
}

void pulseaudio_sink_add_cb(pa_context* c, const pa_sink_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_add(i, is_last, userdata, TRUE);
}

void pulseaudio_sink_change_cb(pa_context* c, const pa_sink_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_add(i, is_last, userdata, FALSE);
}

void pulseaudio_sink_add(const pa_sink_info* i, int is_last, void* userdata, gboolean is_new)
{
    if(is_last < 0)
    {
        g_message("Failed to get sink information: %s",
                pa_strerror(pa_context_errno(context)));
        return;
    }

    if(is_last)
        return;

    if(is_new)
    {
        gchar* msg = g_strdup_printf("new sink \"%s\"", i->description);
        notify(msg, i->name, NULL);
        g_free(msg);
    }

    menu_info_t* mi = userdata;
    char* tooltip = sink_info_str(i);
    menu_info_item_update(mi, i->index, i->name, i->description, &i->volume,
            i->mute, tooltip, NULL, NULL, -1);
    g_free(tooltip);
}

void pulseaudio_source_init_cb(pa_context* c, const pa_source_info* i, int is_last, void* userdata)
{
    pulseaudio_source_add(i, is_last, userdata, FALSE);
}

void pulseaudio_source_add_cb(pa_context* c, const pa_source_info* i, int is_last, void* userdata)
{
    pulseaudio_source_add(i, is_last, userdata, TRUE);
}

void pulseaudio_source_change_cb(pa_context* c, const pa_source_info* i, int is_last, void* userdata)
{
    pulseaudio_source_add(i, is_last, userdata, FALSE);
}

void pulseaudio_source_add(const pa_source_info* i, int is_last, void* userdata, gboolean is_new)
{
    if(is_last < 0)
    {
        g_message("Failed to get source information: %s", pa_strerror(pa_context_errno(context)));
        return;
    }

    if(is_last)
        return;

    const char* class = pa_proplist_gets(i->proplist, PA_PROP_DEVICE_CLASS);

    // ignore monitors
    if(class && g_str_equal(class, "monitor"))
        return;

    if(is_new)
    {
        gchar* msg = g_strdup_printf("new source \"%s\"", i->description);
        notify(msg, i->name, NULL);
        g_free(msg);
    }

    menu_info_t* mi = userdata;
    char* tooltip = source_info_str(i);
    menu_info_item_update(mi, i->index, i->name, i->description, &i->volume,
            i->mute, tooltip, NULL, NULL, -1);
    g_free(tooltip);
}

void pulseaudio_sink_input_init_cb(pa_context* c, const pa_sink_input_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_input_add(i, is_last, userdata, FALSE);
}

void pulseaudio_sink_input_add_cb(pa_context* c, const pa_sink_input_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_input_add(i, is_last, userdata, TRUE);
}

void pulseaudio_sink_input_change_cb(pa_context* c, const pa_sink_input_info* i, int is_last, void* userdata)
{
    pulseaudio_sink_input_add(i, is_last, userdata, FALSE);
}

void pulseaudio_sink_input_add(const pa_sink_input_info* i, int is_last, void* userdata, gboolean is_new)
{
    if(is_last < 0)
    {
        g_message("Failed to get sink information: %s", pa_strerror(pa_context_errno(context)));
        return;
    }

    if(is_last)
        return;

    // ignore filters
    const char* media_role = pa_proplist_gets(i->proplist, PA_PROP_MEDIA_ROLE);
    if(media_role && g_str_equal(media_role, "filter"))
        return;

    const char* app_name = pa_proplist_gets(i->proplist, PA_PROP_APPLICATION_NAME);
    const char* app_icon = pa_proplist_gets(i->proplist, PA_PROP_APPLICATION_ICON_NAME);

    menu_info_t* mi = userdata;
    char* tooltip = input_info_str(i);
    menu_info_item_update(mi, i->index, NULL, app_name ? app_name : i->name,
            &i->volume, i->mute, tooltip, app_icon, NULL, i->sink);
    g_free(tooltip);
}

void pulseaudio_source_output_init_cb(pa_context* c, const pa_source_output_info* i, int is_last, void* userdata)
{
    pulseaudio_source_output_add(i, is_last, userdata, FALSE);
}

void pulseaudio_source_output_add_cb(pa_context* c, const pa_source_output_info* i, int is_last, void* userdata)
{
    pulseaudio_source_output_add(i, is_last, userdata, TRUE);
}

void pulseaudio_source_output_change_cb(pa_context* c, const pa_source_output_info* i, int is_last, void* userdata)
{
    pulseaudio_source_output_add(i, is_last, userdata, FALSE);
}

void pulseaudio_source_output_add(const pa_source_output_info* i, int is_last, void* userdata, gboolean is_new)
{
    if(is_last < 0)
    {
        g_message("Failed to get source information: %s", pa_strerror(pa_context_errno(context)));
        return;
    }

    if(is_last)
        return;

    // ignore peak detects
    const char* media_name = pa_proplist_gets(i->proplist, PA_PROP_MEDIA_NAME);
    if(media_name && g_str_equal(media_name, "Peak detect"))
        return;

    const char* app_name = pa_proplist_gets(i->proplist, PA_PROP_APPLICATION_NAME);
    const char* app_icon = pa_proplist_gets(i->proplist, PA_PROP_APPLICATION_ICON_NAME);

    menu_info_t* mi = userdata;
    char* tooltip = output_info_str(i);
    menu_info_item_update(mi, i->index, NULL, app_name ? app_name : i->name,
            &i->volume, i->mute, tooltip, app_icon, NULL, i->source);
    g_free(tooltip);
}

void pulseaudio_quit(const char* msg)
{
    if(msg)
        g_message("%s", msg);

    if(context)
        pa_context_unref(context);

    if(m)
    {
        pa_signal_done();
        pa_glib_mainloop_free(m);
    }
}
